cpp/src/arrow/compute/kernels/util-internal.h

```diff
@@ -47,11 +52,11 @@ using enable_if_timestamp =
     typename std::enable_if<std::is_base_of<TimestampType, T>::value>::type;
 
 template <typename T>
-using enable_if_has_c_type =
-    typename std::enable_if<std::is_base_of<PrimitiveCType, T>::value ||
-                            std::is_base_of<DateType, T>::value ||
-                            std::is_base_of<TimeType, T>::value ||
-                            std::is_base_of<TimestampType, T>::value>::type;
+using enable_if_has_c_type = typename std::enable_if<
+    !std::is_same<UInt8Type, T>::value && !std::is_same<Int8Type, T>::value &&
```

Comment:
It would be better to exclude the 8-bit integers in the functor declarations in the hash module, adding this here makes the definition misleading (since int8, uint8 have a `c_type`)

Reply:
I am not sure how to exclude 8-bit integers in the functor declarations. I am thinking creating a new template function like "enable_if_has_c_type_excluding_8bit_integer" here and use it in the declaration of hash table pass for primitive types